### DISCUSSION TITLE
Integrating with prisma with support for local development and RLS

### DISCUSSION THREAD

**Discussion by NixBiks** [supabase/supabase#4051] [Questions]
I like Prisma and I like supabase! I have a pretty good project where they coexist but there are few obstacles that I'd like to discuss how people tackle. The real challenge is to enable RLS in a local setup. 

## Deploy local supabase

With supabase cli it's really easy to get setup locally. You do `supabase init` and `supabase start`.

## Migrations

I like to control all database migrations with `prisma migrate` but here we have one challenge. Prisma does not [support querying across multiple schemas](https://github.com/prisma/prisma/issues/1122) so for now we can only use prisma migrate on operations in our `public` schema. So my first question is; how do people handle migrations in other schemas, e.g. `auth`, in a similar setup? The need for this becomes clear in the next section.

## Row level security

Preferably we'd have relations to `auth.users` table to easily exploit RLS. Since we can't do that with our prisma model we can do something similar - reflect `auth.users` in `public."Users"`. To achieve that I can add the following to `schema.prisma`  

```prisma
model Users {
  id    String @id
  email String @unique
}
```

and create some postgres triggers to sync

```sql

-- inserts a row into public."Users"
create or replace function public.handle_new_user() 
returns trigger as $$
begin
  insert into public."Users" (id, email)
  values (new.id, new.email);
  return new;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is created
drop trigger if exists on_auth_user_created on auth.users;
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();


-- update a row in public."Users" when the email is updated
create or replace function public.handle_updated_user()
returns trigger as $$
begin
  update public."Users"
  set email = new.email
  where id = new.id::text;
  return new;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is updated
drop trigger if exists on_auth_user_updated on auth.users;
create trigger on_auth_user_updated
  after update on auth.users
  for each row execute procedure public.handle_updated_user();


-- delete a row from public."Users" when the user is deleted
create or replace function public.handle_deleted_user()
returns trigger as $$
begin
  delete from public."Users" where id = old.id::text;
  return old;
end;
$$ language plpgsql security definer;

-- trigger the function every time a user is deleted
drop trigger if exists on_auth_user_deleted on auth.users;
create trigger on_auth_user_deleted
  after delete on auth.users
  for each row execute procedure public.handle_deleted_user();
```

So this is great - there is a "1-1 relationship" on `public.Users` and `auth.users` but how can I migrate the triggers? I can't add them to my prisma migrations because there are references to `auth` schema.

I'm also considering if I'm getting off track here - maybe there is a better way to implement RLS with Prisma and supabase auth? Maybe something like [this](https://github.com/prisma/prisma/issues/5128#issuecomment-915815199)? But it isn't clear to me exactly how that should work.

**Comment by xpressui** [comment]
The way I work around this is by moving the auth schema related stuff out of the migration and put them in their own sql file. Then I write a script that applies the prisma migrations first and the auth stuff last (using psql -f).

**Reply by xpressui** [reply]
also, if you plan to use the auth functions in your migrations (auth.uid(), auth.email(), auth.role()), you can create placeholders for them in public schema (i.e public.auth_uid(), public.auth_email(), etc) and use those instead (they can return just dummy data). You can then replace their dummy implementation with calls to their auth equivalents (using create or replace function) after the migration is done.

**Reply by dsebastien** [reply]
Sound like a clean approach for now. Do you have an example?

**Reply by dsebastien** [reply]
I've created these for now, but I have yet to test/validate. I'll try to report back later:

```
create or replace function public.auth_email()
  returns text as $$
begin
return auth.email();
end;
$$ language plpgsql;

create or replace function public.auth_jwt()
  returns text as $$
begin
return auth.jwt();
end;
$$ language plpgsql;

create or replace function public.auth_role()
  returns text as $$
begin
return auth.role();
end;
$$ language plpgsql;

create or replace function public.auth_uid()
  returns text as $$
begin
return auth.uid();
end;
$$ language plpgsql;

```

**Reply by maxcountryman** [reply]
@dsebastien are you running that in a script as part of the Prisma migration?

**Reply by dsebastien** [reply]
I've added that to a .sql file within the project and I import it in a "preseed" script with psql. Thus deployed once when initializing the environment

**Comment by drkstr101** [comment]
Is there a way to get prisma to use the auth context from supabase auth? My nieve understanding is that prisma usage will bypass all row-level security, running the queries as `postgres` user. Is my understanding incorrect? I wanted to use prisma in my latest project, but I don't want to lose out on RLS. Ideas anyone?

Cheers!

**Reply by xpressui** [reply]
This is one way to do it: https://github.com/prisma/prisma/issues/5128

**Comment by solvionsolutions** [comment]
Yeah, your instinct is right — with the default `postgres` connection string those queries never touch RLS at all. That role owns the tables and carries `BYPASSRLS`, so Prisma is effectively running as a superuser and your policies just don't fire, no matter what you do with the `public."Users"` mirror. Two things fix it: point `DATABASE_URL` at a dedicated login role that isn't the table owner and has `NOBYPASSRLS`, then wrap each request in a transaction where you switch into `authenticated` and inject the JWT claims so `auth.uid()` actually resolves.

```ts
await prisma.$transaction(async (tx) => {
  await tx.$executeRaw`select set_config('request.jwt.claims', ${JSON.stringify(claims)}, true)`;
  await tx.$executeRaw`set local role authenticated`;
  // queries in here now run under RLS, with auth.uid() available
});
```

Because it's `set local`, the role and claims reset when the transaction commits, so nothing leaks between requests. Switching to the restricted role is also the fastest way to prove the policies were being skipped before — you'll suddenly get empty results where the same query used to hand back every row. Even if you've long since moved past this, it trips up a lot of people who land here from search, so figured it was worth spelling out.

Full breakdown of why ORMs slip past Supabase RLS: https://www.guardlayer.io/blog/orm-bypasses-supabase-rls
